### DISCUSSION TITLE
帮助解释一下README.md中关于透明代理client和正向代理client的一段话

### DISCUSSION THREAD

**Discussion by fxchby** [zfl9/ss-tproxy#219] [Q&A]
在README.me中有这些一句话：
……………………
另外还有一点需要注意，透明代理使用的 client 与正向代理使用的 client 通常是不同的，_因为正向代理的 client 是 http、socks5 服务器，而透明代理的 client 则是透明代理服务器，_它们之间有本质上的区别。
……………………
其中的：**因为正向代理的 client 是 http、socks5 服务器，而透明代理的 client 则是透明代理服务器**
这句话不理解。
能帮忙详细解释一下
另外，透明代理服务器这种说法，是不是也有点含糊。代理服务器本身不可能做到透明，必须有其它配置才行。
是我的理解有错？请帮忙指出

**Comment by cattyhouse** [comment]
你可以将此处的 `client` 理解为 `拦截流量的中间人`

**Comment by fxchby** [comment]
明白了，谢谢
这里的client应该总是相对于远程代理服务器而言的，指代理client

**Comment by zfl9** [comment]
这里的 client/server 指 代理套件 中的 client/server 组件。比如 shadowsocks-libev：

- ss-server 是 "server"
- ss-local 是 传入协议为 socks5 的 "client"
- ss-redir 是 传入协议为 redirect/tproxy 的 "client"
- ss-tunnel 也可以看作是一种 "client"，其转发目标是固定的 ip:port，因此传入协议取决于目标 ip:port

> 可以将此处的 "client" 理解为 "agent"，本地流量都是经过这些 "agent" 传输至 "server" 的

至于 "透明代理服务器"，其实我的意思是说 "传入协议为 redirect/tproxy 的 本地服务器"，即 ss-redir

之所以它是 "本地服务器"，是因为 ss-redir 和 ss-local 类似，都会在本地 bind 一个 socket 端口，这个端口提供一个 "代理服务"，这符合 "服务器" 的定义。他们的区别只是 "协议" 不同，要与 ss-local 通信，用的是 socks5 协议，要与 ss-redir 通信，用的是 redirect/tproxy 协议。

或者你也可以这样理解，从 "走了代理的APP" 的角度看，ss-local/ss-redir 就是他们的 "代理服务器"，对于这些 APP，他们其实不知道（也不需要知道）藏在 ss-local/ss-redir 后面的 ss-server。

> 此处将 redirect/tproxy 描述为一种"协议"不是很严谨，严格来说它们不是"网络协议"。但为了便于理解，请允许我这样做。

**Comment by zfl9** [comment]
因为这里实际上涉及了 3 个角色：

 - A: 走代理的APP/进程，如 curl
 - B: 本地的代理服务器，如 ss-local/ss-redir
 - C: 远程的代理服务器，如 ss-server

--- 

- A,B 是一对 client/server
- B,C 也是一对 client/server

---

因此当你看到 client/server、客户端/服务器 等术语时，你需要了解当前处于什么"上下文"。

**Comment by fxchby** [comment]
明白了，非常详细，谢谢大佬
